### PR TITLE
feat: Add Johnson's algorithm

### PR DESCRIPTION
**Pull request by starovoid**
## Brief
Add [Johnson algorithm][johnson] for all pairs shortest path problem.

```rust
pub fn johnson<G, F, K>(
    graph: G,
    mut edge_cost: F,
) -> Result<HashMap<(G::NodeId, G::NodeId), K>, NegativeCycle>
where
    G: IntoEdges + IntoNodeIdentifiers + NodeIndexable + Visitable,
    G::NodeId: Eq + Hash,
    F: FnMut(G::EdgeRef) -> K,
    K: BoundedMeasure + Copy + Sub<K, Output = K>,
```

## About algorithm & implementation
Johnson's algorithm is based on Dijkstra's algorithm, calling it for each vertex after some preparation using SPFA.
In this way we can find all pair shortest paths in O(VD + VE) time (D is the Dijksta's time complexity).

In the current implementation, we have a time complexity O(VE*log(V) + V^2*log(V)), which is preferable for non-dense graphs, compared to the Floyd-Warshall algorithm.

Instead of SPFA, it would also be possible to use the existing `bellman_ford` implementation, but SPFA usage additionally speeds up the algorithm on sparse graphs, and there is an unconfirmed estimate that SPFA works on average in O(E) time.

## Becnhmarks
I have conducted small comparative tests that confirm the statement that Johnson's algorithm is preferable on sparse graphs (almost five times faster than `floyd_warshall`), but it also naturally loses on dense graphs.

```
test johnson_dense_1000_nodes  ... bench: 1,169,255,997.40 ns/iter (+/- 7,208,144.98)
test johnson_dense_100_nodes   ... bench:   1,777,834.38 ns/iter (+/- 24,188.31)
test johnson_sparse_1000_nodes ... bench:  86,516,820.50 ns/iter (+/- 1,082,174.18)
test johnson_sparse_100_nodes  ... bench:     778,894.19 ns/iter (+/- 14,603.94)
```

```
test floyd_warshall_dense_1000_nodes  ... bench: 544,423,954.20 ns/iter (+/- 7,272,191.85)
test floyd_warshall_dense_100_nodes   ... bench:     568,021.70 ns/iter (+/- 12,867.89)
test floyd_warshall_sparse_1000_nodes ... bench: 495,226,815.90 ns/iter (+/- 8,736,234.74)
test floyd_warshall_sparse_100_nodes  ... bench:     529,623.65 ns/iter (+/- 14,083.83)

```
The benchmark code has also been added to this PR in order to be able to compare the performance of certain algorithms in the future.

## `parallel_johnson`
A version of the algorithm that parallels `dijkstra` calls is available under the `rayon` feature.

```rust
pub fn parallel_johnson<G, F, K>(
    graph: G,
    mut edge_cost: F,
) -> Result<HashMap<(G::NodeId, G::NodeId), K>, NegativeCycle>
where
    G: IntoEdges + IntoNodeIdentifiers + NodeIndexable + Visitable + Sync,
    G::NodeId: Eq + Hash + Send,
    F: Fn(G::EdgeRef) -> K + Sync,
    K: BoundedMeasure + Copy + Sub<K, Output = K> + Send + Sync,
```

The benchmarks show the following performance on PC with Ryzen 9 7950X:
```
test johnson_dense_1000_nodes           ... bench: 1,177,432,008.20 ns/iter (+/- 11,077,709.38)
test johnson_dense_100_nodes            ... bench:   1,714,581.45 ns/iter (+/- 31,400.66)
test johnson_sparse_1000_nodes          ... bench:  78,071,539.40 ns/iter (+/- 2,063,005.61)
test johnson_sparse_100_nodes           ... bench:     660,780.69 ns/iter (+/- 15,389.23)
test parallel_johnson_dense_1000_nodes  ... bench:  68,611,940.20 ns/iter (+/- 1,782,308.55)
test parallel_johnson_dense_100_nodes   ... bench:     258,762.17 ns/iter (+/- 16,887.11)
test parallel_johnson_sparse_1000_nodes ... bench:  14,309,925.60 ns/iter (+/- 1,388,024.20)
test parallel_johnson_sparse_100_nodes  ... bench:     259,247.53 ns/iter (+/- 15,763.51)
```

and
```
test johnson_dense_1000_nodes           ... bench: 1,931,386,001.60 ns/iter (+/- 30,692,353.78)
test johnson_sparse_1000_nodes          ... bench: 155,685,943.60 ns/iter (+/- 104,165,471.95)
test parallel_johnson_dense_1000_nodes  ... bench: 246,817,415.40 ns/iter (+/- 7,077,236.51)
test parallel_johnson_sparse_1000_nodes ... bench:  34,137,779.80 ns/iter (+/- 4,018,118.69)
```
on laptop with 12th Intel(R) i7-12700H

[johnson]: https://en.wikipedia.org/wiki/Johnson%27s_algorithm

### CODE REVIEW

File: src/algo/johnson.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    let mut distance_map: HashMap<(G::NodeId, G::NodeId), K> =
+        HashMap::with_capacity(node_bound * node_bound);
+
+    // Reweight edges.
+    let mut new_cost = |edge: G::EdgeRef| {
+        let (sum, _overflow) = edge_cost(edge).overflowing_add(reweight[ix(edge.source())]);
+        debug_assert!(!_overflow);
+        sum - reweight[ix(edge.target())]
+    };
```
Comment (RaoulLuque) [inline src/algo/johnson.rs:109]: Maybe, we could add a check that tests whether there actually are any negative weight edges? Because if there aren't, most of these steps could be skipped which might safe quite some time?
This might mean imposing more trait bounds on K, or might not even. One could use K::default() from `BoundedMeasure` (technically, `Measure`), which should be 0 I guess? Or there might be a more safe approach.

**Review by RaoulLuque — COMMENTED** [review]
Seems like a useful addition to Petgraph :)

Was wondering a few things:

I saw that this is similar with Floyd Warshall, but why do we not return the actual paths that are computed in some way? Of course changing this for Floyd Warshall would be an API change, but we could do it from the start with Johnson's?

Probably this would be a performance drawback, but would it maybe be possible to discard the "path map" immediately as a user in which case the compiler might optimize the performance slowdown away?
I.e. something like `let (res, _) = johnson(...).unwrap()`, where the `_` would be throwing away the "path map" returned by johnson. Not sure if this works, especially with the Result return type, but might be worth looking into 🤔

**Comment by starovoid** [comment]
> but why do we not return the actual paths that are computed in some way? Of course changing this for Floyd Warshall would be an API change, but we could do it from the start with Johnson's?

Because it will also require changing/adding a new version of `dijkstra`, which is beyond the scope of a single PR. 

Secondly, significant refactoring and generalization of shortest path algorithms is expected in the future, and then I hope we will be able to develop a uniform API for both versions of all algorithms without paying extra for unused functionality. Some work has already been done previously in the `next` branch, and I expect it to resume in the coming months.

File: benches/floyd_warshall.rs [hunk 2]
```diff
@@ -31,3 +31,43 @@ fn floyd_warshall_bench(bench: &mut Bencher) {
...
+fn floyd_warshall_dense_1000_nodes(bench: &mut Bencher) {
+    let graph = build_graph(1000, true);
+
+    bench.iter(|| {
+        let _scores = floyd_warshall(&graph, |e| *e.weight());
+    });
+}
+
+#[allow(clippy::needless_range_loop)]
+fn build_graph(node_count: usize, dense: bool) -> Graph<usize, i32, Undirected> {
```
[the remark this answers is not in the document]
Comment (starovoid) [reply benches/floyd_warshall.rs:54]: The graph is generated using a specific algorithm ("popular" among our benchmarks), and it's just easier that way. A major rewrite of benchmarks is also expected in the `next` branch.

File: src/algo/johnson.rs (hunk 1 above)
Comment (starovoid) [reply src/algo/johnson.rs:109]: I think such a check would be unnecessary, since for a graph with obviously non-negative weights, the user would simply prefer to call `dijkstra` `n` times.

**Comment by RaoulLuque** [comment]
> Because it will also require changing/adding a new version of `dijkstra`, which is beyond the scope of a single PR.
> 
> Secondly, significant refactoring and generalization of shortest path algorithms is expected in the future, and then I hope we will be able to develop a uniform API for both versions of all algorithms without paying extra for unused functionality. Some work has already been done previously in the `next` branch, and I expect it to resume in the coming months.

Oh, I didn't see that `dijkstra`'s API didn't support that either. Of course one could change the actual functionality behind the `dijkstra` API and expose a new one also returning the paths and let the old one simply call the new one throwing away the paths. But I guess in that case it would make sense to wait for these changes for the `next` branch. Are there any time estimates on when changes in that branch might land?

File: benches/floyd_warshall.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RaoulLuque) [reply benches/floyd_warshall.rs:54]: Okay, that's nice that there will be a rewrite. Until then tho, it seems useful to know how the graph roughly looks like and the docstring could be quite brief I would think

File: src/algo/johnson.rs (hunk 1 above)
Comment (RaoulLuque) [reply src/algo/johnson.rs:109]: Not sure about that. Since Floyd Warshall actually does something a bit different for example than Dijkstra, users might think that it is not worth it running Dijkstra from every possible starting graph. I mean, users of the Johnsons algorithm wouldn't necessarily know that that's actually what's going on under the hood.

I think it would be worth testing first, whether there would actually be a noticeable speedup or not, and if there is, one could still add a comment about that to the docstring to less graph-algorithm-knowledgeable users know of that option.

**Comment by starovoid** [comment]
> Are there any time estimates on when changes from that branch might land in main?

Work on `next` is awaiting the author's return (1-2 months), then I plan to join there when `petgraph 0.8.3` released. But it's difficult to give any estimates - there are a lot of tasks there.

File: benches/johnson.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    bench.iter(|| {
+        let _scores = johnson(&g, |e| *e.weight());
+    });
+}
+
+#[bench]
+#[cfg(feature = "rayon")]
+fn parallel_johnson_bench(bench: &mut Bencher) {
```
[the remark this answers is not in the document]
Comment (starovoid) [reply benches/johnson.rs]: I've moved the test graph generation code to `banches/common`, now it should be more clear.

File: src/algo/johnson.rs (hunk 1 above)
Comment (starovoid) [reply src/algo/johnson.rs:109]: I prefer the option with a comment. It's better to help the user become a little more understanding than to have an overhead when it's not needed.

File: src/algo/floyd_warshall.rs [hunk 4]
```diff
@@ -21,6 +21,9 @@ use crate::visit::{
 /// * `Ok`: (if graph contains no negative cycle) a hashmap containing all pairs shortest paths
 /// * `Err`: if graph contains negative cycle.
 ///
+/// **Note**: If the graph is sparse (the number of edges is quite small),
+/// consider using the [`johnson`](fn@crate::algo::johnson), which is likely to show better performance.
+///
```
[the remark this answers is not in the document]
Comment (starovoid) [reply src/algo/floyd_warshall.rs:26]: This would be a pretty bold statement, since in addition to the absolute operating time, resource consumption is often also taken into account, which will be much more economical when using `floyd_warshall` for dense graphs.

File: benches/floyd_warshall.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (starovoid) [reply benches/floyd_warshall.rs]: I haven't really figured out how to describe this graph briefly. "Well, edge сount..., what is there to talk about?"

[the remark this answers is not in the document]
Comment (RaoulLuque) [reply benches/floyd_warshall.rs]: I see,  yeah. I thought there might be a closed formula for the edge count or the resulting graph would have a certain shape but if that's not the case there is not a lot to say, true.

File: src/algo/floyd_warshall.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (RaoulLuque) [reply src/algo/floyd_warshall.rs:26]: You are definitely right that one has to take into account the resource consumption, but it still might be interesting to users, to know that parallelized johnson might run faster (not necessarily more efficient) than normal floyd warshall, even on general graphs. But I also see your point, so feel free to include a comment about the runtime/resource trade-off or not 👍

File: src/algo/johnson.rs (hunk 1 above)
Comment (RaoulLuque) [reply src/algo/johnson.rs:112]: Yeah, makes sense. Seems like a good compromise :)
